float32[4] inputs